### PR TITLE
feat: Dynamic dashboard component

### PR DESCRIPTION
**Pull request by simcha90**
### SUMMARY
This PR is the implementation of SIP -76 #17210 - feat for dynamic loading of dashboard components

**Why?**

Dependent on product needs we want to take development of Superset to different directions. `Dashboard Components` can introduce specific needs of company requirements. So we wanted to do this feature more flexible in order to not affect other product visions :)

**What?**

This PR add opportunity to add new `Dashboard Components` that can be configured in separate file by every superset instance and not affects main Superset functionality. Commonly needed types relating to the Dashboard are also moved into `@superset-ui/core` to make it possible to fully decouple dashboard component plugin development from the Superset codebase.

**How?**

We creating new dashboard components registry that can register new `Dashboard Components`  in `/src/setup/setupDasboardComponents.ts`. In a follow-up PR, this code will be moved into a Yeoman generator template.

When we drag new component to dashboard we use `React.lazy` function to load code of component

**NOTES:**
- We created new type of registry using functional approach in favor of  class approach that we are using now for loading plugin charts
- We have some basic predefined rules for dynamic components
- To test existing functionality you need uncomment code in file `superset-frontend/src/setup/setupDasboardComponents.ts` and it will add demo component to list of components

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
https://user-images.githubusercontent.com/56388545/138594705-32b5a862-295f-4111-a63b-d8e8c96fcb7e.mov

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/visualizations/dashboardComponents/TestComponent/index.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+// TODO: POC only component can be removed after PR approved
+export default {
+  metadata: {
+    name: 'Test',
+    description: 'Test description',
+    iconName: 'filter',
+  },
+  loadComponent: () => import('./TestComponent'),
```
Comment (villebro) [inline superset-frontend/src/visualizations/dashboardComponents/TestComponent/index.ts:27]: Should we rather call this an "Example component" to disambiguate from regular tests?

**Review by villebro — COMMENTED** [review]
First pass comments. Code looks great, but my main question is shouldn't we move the central building blocks to `@superset-ui/core`, like the central types/interfaces and registries? If we'd do that, we could
1) Develop the plugins independently from the Superset codebase.
2) Add a dashboard component template to the Yeoman generator. See #18027 which was merged just recently and made the system work again after the monorepo migration
3) Publish dashboard component plugins to npm
4) Pin plugins to specific versions etc

Etc. Also, I think we should add tests as early as possible, which are missing here. By moving the relevant parts to `@superset-ui/core` we would automatically have 100 % test coverage requirements, but in addition, we should also have at least a few RTL tests that make sure the relevant Dashboard flows work (viewing a dashboard with a plugin + adding and removing a component)

**Comment by simcha90** [comment]
> First pass comments. Code looks great, but my main question is shouldn't we move the central building blocks to `@superset-ui/core`, like the central types/interfaces and registries? If we'd do that, we could
> 
> 1. Develop the plugins independently from the Superset codebase.
> 2. Add a dashboard component template to the Yeoman generator. See #18027 which was merged just recently and made the system work again after the monorepo migration
> 3. Publish dashboard component plugins to npm
> 4. Pin plugins to specific versions etc
> 
> Etc. Also, I think we should add tests as early as possible, which are missing here. By moving the relevant parts to `@superset-ui/core` we would automatically have 100 % test coverage requirements, but in addition, we should also have at least a few RTL tests that make sure the relevant Dashboard flows work (viewing a dashboard with a plugin + adding and removing a component)

May be functional registry can be moved to `superset-core`, but meantime I don't see some special need for it because:
1. we can do it also now, I think this PR enough for our company needs to start develop our plugins
2. I'm not sure that we have capacity for it, @amitmiran137 do we have?
3, 4 Can be done also now, we don't need move code to `superset-core` for these things

P.S. other stuff fixed

**Review by zhaoyongjie — CHANGES_REQUESTED** [review]
Hi @simcha90, in order to local test, could you rebase this PR from the master branch? Thanks!

I got some lint errors in my local. I think if we rebase master might suppress this issue in IDE.
image

**Comment by simcha90** [comment]
@zhaoyongjie done

**Review by villebro — COMMENTED** [review]
The code LGTM and in general I think the proposed solution looks great. However, to further decouple this and make it more approachable for plugin developers, I'd like to propose the following:
1) add a Yeoman template, similar to the viz plugin one, to `superset-ui/generator-superset`. This would optimally generate the example component. By doing this we would not need the example component in the actual codebase
2) Add a section to the docs on how to create a dashboard plugin, something similar to https://superset.apache.org/docs/contributing/creating-viz

My long term vision is to keep making it easier to add plugins, hopefully removing the need for manually registering the plugins in the `superset-frontend` codebase all together (I'm thinking moving more and more in the direction of dynamic plugins). To spread the pain, I'm happy to tackle the docs, and can even help out moving the example plugin to the yeoman generator if needed.

**Review by villebro — APPROVED** [review]
LGTM

**Review by zhaoyongjie — APPROVED** [review]

**Pull request merged**